File: src/data/caseFiles.ts
import { PuzzleData } from '../types';
import { isMorningEdition, isNightEdition } from '../utils/edition';
import data from './caseFiles.json';

export type CaseSlot = 'Morning' | 'Evening';
export type CaseCharacterId =
  | 'thorne'
  | 'beatrice'
  | 'clara'
  | 'sterling'
  | 'archibald'
  | 'blackwood'
  | 'reginald';

export type CasePart =
  | { kind: 'text'; value: string; when?: CaseSlot }
  | { kind: 'quote'; slot: CaseSlot };

export type CaseFragment = {
  characterId: CaseCharacterId;
  editionNumber: number;
  title: string;
  parts: CasePart[];
};

export type CaseNoteSegment = { kind: 'text' | 'quote'; value: string };

export type AssembledFragment = {
  characterId: CaseCharacterId;
  editionNumber: number;
  title: string;
  segments: CaseNoteSegment[];
};

export type CaseCharacter = {
  id: CaseCharacterId;
  name: string;
  dossier: string;
  file: string;
  plate: string;
};

export const CASE_CHARACTERS: CaseCharacter[] = data.characters as CaseCharacter[];

export const CASE_FRAGMENTS: CaseFragment[] = data.fragments as CaseFragment[];

export function findEditionPuzzle(
  puzzles: PuzzleData[],
  editionNumber: number,
  slot: CaseSlot
) {
  return puzzles.find((puzzle) => {
    if (puzzle.editionNumber !== editionNumber) return false;
    return slot === 'Evening' ? isNightEdition(puzzle) : isMorningEdition(puzzle);
  });
}

function slotSolved(
  puzzles: PuzzleData[],
  solvedPuzzleIds: string[],
  editionNumber: number,
  slot: CaseSlot
) {
  const puzzle = findEditionPuzzle(puzzles, editionNumber, slot);
  return Boolean(puzzle && solvedPuzzleIds.includes(puzzle.id));
}

function partVisible(
  part: CasePart,
  puzzles: PuzzleData[],
  solvedPuzzleIds: string[],
  editionNumber: number
) {
  if (part.kind === 'quote') {
    return slotSolved(puzzles, solvedPuzzleIds, editionNumber, part.slot);
  }
  if (!part.when) return true;
  return slotSolved(puzzles, solvedPuzzleIds, editionNumber, part.when);
}

function pushText(segments: CaseNoteSegment[], value: string) {
  const last = segments[segments.length - 1];
  if (last?.kind === 'quote') {
    const ended = /[.!?]$/.test(last.value);
    value = value.replace(/^\s*[.]+(?=\s|$)/, '');
    const next = value.trim();
    if (!next) return;
    if (ended && (/^[,;:]/.test(next) || /^[a-z]/.test(next))) {
      last.value = last.value.replace(/[.!?]$/, '');
    }
    value = /^[,:;]/.test(next) ? next : ` ${next}`;
  }
  if (!value.trim()) return;
  segments.push({ kind: 'text', value });
}

export function assembleFragment(
  fragment: CaseFragment,
  puzzles: PuzzleData[],
  solvedPuzzleIds: string[]
): AssembledFragment | null {
  const segments: CaseNoteSegment[] = [];
  let hasQuote = false;

  for (const part of fragment.parts) {
    if (!partVisible(part, puzzles, solvedPuzzleIds, fragment.editionNumber)) continue;
    if (part.kind === 'quote') {
      const puzzle = findEditionPuzzle(puzzles, fragment.editionNumber, part.slot);
      if (!puzzle) continue;
      segments.push({
        kind: 'quote',
        value: puzzle.originalText,
      });
      hasQuote = true;
      continue;
    }
    pushText(segments, part.value);
  }

  if (!hasQuote) return null;
  return {
    characterId: fragment.characterId,
    editionNumber: fragment.editionNumber,
    title: fragment.title,
    segments,
  };
}

export function unlockedFragmentsForCharacter(
  characterId: CaseCharacterId,
  puzzles: PuzzleData[],
  solvedPuzzleIds: string[]
) {
  return CASE_FRAGMENTS.filter((fragment) => fragment.characterId === characterId)
    .map((fragment) => assembleFragment(fragment, puzzles, solvedPuzzleIds))
    .filter((fragment): fragment is AssembledFragment => Boolean(fragment));
}

export function hasDecodedFragments(puzzles: PuzzleData[], solvedPuzzleIds: string[]) {
  return CASE_FRAGMENTS.some((fragment) => assembleFragment(fragment, puzzles, solvedPuzzleIds));
}

export function fragmentKey(fragment: { characterId: string; editionNumber: number }) {
  return `${fragment.characterId}-${fragment.editionNumber}`;
}

export function fragmentsUpdatedByPuzzle(puzzle: PuzzleData) {
  const slot: CaseSlot = isNightEdition(puzzle) ? 'Evening' : 'Morning';
  return CASE_FRAGMENTS.filter(
    (fragment) =>
      fragment.editionNumber === puzzle.editionNumber &&
      fragment.parts.some((part) => part.kind === 'quote' && part.slot === slot)
  );
}

File: android/core/content/src/main/kotlin/com/chroniclecryptogram/content/CaseFiles.kt
package com.chroniclecryptogram.content

import com.chroniclecryptogram.cipher.Edition
import com.chroniclecryptogram.cipher.model.PuzzleData

/** One assembled line of a dossier: prose, or a quote the player decoded. */
data class CaseNoteSegment(val kind: Kind, val value: String) {
    enum class Kind { TEXT, QUOTE }
}

data class AssembledFragment(
    val characterId: String,
    val editionNumber: Int,
    val title: String,
    val segments: List<CaseNoteSegment>,
)

/**
 * Assembles case-file dossiers from what the player has actually solved, ported
 * from the behaviour half of `src/data/caseFiles.ts` and pinned by the
 * case-files fixture.
 *
 * The rule that gives the feature its point: a `quote` part splices in the
 * decoded quote for its slot, and a `text` part carrying a `when` appears only
 * once that slot is solved. **A fragment with no quote at all is withheld
 * entirely** -- otherwise the dossier would read as a table of contents for
 * editions the player has not reached.
 */
object CaseFiles {

    private const val MORNING = "Morning"
    private const val EVENING = "Evening"

    private val endsInStop = Regex("[.!?]$")
    private val leadingPeriods = Regex("^\\s*[.]+(?=\\s|$)")
    private val startsWithClause = Regex("^[,;:]")
    private val startsWithLower = Regex("^[a-z]")
    private val startsWithPunctuation = Regex("^[,:;]")

    fun assemble(
        fragment: CaseFragment,
        puzzles: List<PuzzleData>,
        solvedPuzzleIds: Set<String>,
    ): AssembledFragment? {
        val segments = ArrayList<CaseNoteSegment>()
        var hasQuote = false

        for (part in fragment.parts) {
            if (!partVisible(part, fragment.editionNumber, puzzles, solvedPuzzleIds)) continue

            if (part.kind == "quote") {
                val slot = part.slot ?: continue
                val puzzle = puzzleForSlot(fragment.editionNumber, slot, puzzles) ?: continue
                segments += CaseNoteSegment(CaseNoteSegment.Kind.QUOTE, puzzle.originalText)
                hasQuote = true
                continue
            }

            pushText(segments, part.value.orEmpty())
        }

        // No decoded quote means nothing was earned, so the fragment is withheld.
        if (!hasQuote) return null

        return AssembledFragment(
            characterId = fragment.characterId,
            editionNumber = fragment.editionNumber,
            title = fragment.title,
            segments = segments,
        )
    }

    fun unlockedFragmentsForCharacter(
        characterId: String,
        content: CaseFileContent,
        puzzles: List<PuzzleData>,
        solvedPuzzleIds: Set<String>,
    ): List<AssembledFragment> = content.fragments
        .filter { it.characterId == characterId }
        .mapNotNull { assemble(it, puzzles, solvedPuzzleIds) }

    fun hasDecodedFragments(
        content: CaseFileContent,
        puzzles: List<PuzzleData>,
        solvedPuzzleIds: Set<String>,
    ): Boolean = content.fragments.any { assemble(it, puzzles, solvedPuzzleIds) != null }

    /**
     * Appends prose, stitching the seam where it follows a decoded quote.
     *
     * The authored text assumes the quote ends the sentence, so a continuation
     * needs the quote's full stop removed and the stray leading period dropped.
     * Without this the dossier reads ". , and he paid it." instead of
     * "..., and he paid it."
     */
    private fun pushText(segments: MutableList<CaseNoteSegment>, raw: String) {
        var value = raw
        val last = segments.lastOrNull()

        if (last != null && last.kind == CaseNoteSegment.Kind.QUOTE) {
            val ended = endsInStop.containsMatchIn(last.value)
            value = leadingPeriods.replace(value, "")
            val next = value.trim()
            if (next.isEmpty()) return
            if (ended && (startsWithClause.containsMatchIn(next) || startsWithLower.containsMatchIn(next))) {
                segments[segments.lastIndex] = last.copy(
                    value = endsInStop.replace(last.value, "")
                )
            }
            value = if (startsWithPunctuation.containsMatchIn(next)) next else " $next"
        }

        if (value.isBlank()) return
        segments += CaseNoteSegment(CaseNoteSegment.Kind.TEXT, value)
    }

    private fun partVisible(
        part: CasePart,
        edition: Int,
        puzzles: List<PuzzleData>,
        solvedPuzzleIds: Set<String>,
    ): Boolean {
        if (part.kind == "quote") {
            val slot = part.slot ?: return false
            return slotSolved(edition, slot, puzzles, solvedPuzzleIds)
        }
        val gate = part.`when` ?: return true
        return slotSolved(edition, gate, puzzles, solvedPuzzleIds)
    }

    private fun puzzleForSlot(edition: Int, slot: String, puzzles: List<PuzzleData>): PuzzleData? =
        when (slot) {
            EVENING -> Edition.nightPuzzleForEdition(puzzles, edition)
            MORNING -> Edition.morningPuzzleForEdition(puzzles, edition)
            else -> null
        }

    private fun slotSolved(
        edition: Int,
        slot: String,
        puzzles: List<PuzzleData>,
        solvedPuzzleIds: Set<String>,
    ): Boolean = puzzleForSlot(edition, slot, puzzles)?.id in solvedPuzzleIds
}
